CMS-Enterprise/sbom-harbor-ui
# feat: implement oidc login

## Summary

This PR integrates Cognito federated sign in with CMS IDM into the sbom-harbor-ui project. This new feature allows users to log in with OIDC credentials. A button has been added to the SignIn view which, upon interaction, redirects the user to the OIDC provider page.

### Added

1.  Integration of Cognito federated sign in with CMS IDM.
2.  A new button on the SignIn view for OIDC login.
3.  Implementation of `loginUserFederated` action to facilitate OIDC login.

### Changed

1.  Updates to `SignIn.tsx` to accommodate the new OIDC login button.
2.  Addition of OIDC login functionality in `loginUserFederated.tsx`.
3.  Configuration for Cognito in `configureCognito.ts`.

### Deprecated

- n/a

### Removed

- n/a

### Fixed

- n/a

## How to test

1. Run `yarn lint`, `yarn test`, and `yarn build` to ensure that the codebase is stable and the tests are passing
2. Edit your `.env.development.local` file to contain the newest environment variables in `.env.test` with the correct values.
3. Start the app with `yarn start`
4. Navigate to `http://localhost:5173/#/login` locate the new button for OIDC login.
5. Click on the OIDC login button. This should redirect you to the OIDC provider page.
6. Try logging in with valid OIDC credentials. The login should be successful and redirect you back to the application.
7. Try logging in with invalid credentials. The login should fail with an appropriate error message.
8. Verify that other functionalities of the application remain unaffected by this new integration.

### Test Coverage

```
-------------------------------------|---------|----------|---------|---------|---------------------------
File                                 | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
-------------------------------------|---------|----------|---------|---------|---------------------------
All files                            |   38.52 |    22.45 |   16.46 |   34.71 |
 src                                 |   74.13 |    88.88 |   57.14 |   38.09 |
  main.tsx                           |       0 |        0 |       0 |       0 | 14-45
  types.ts                           |     100 |      100 |     100 |     100 |
 src/actions                         |     100 |      100 |     100 |     100 |
  actionTypes.ts                     |     100 |      100 |     100 |     100 |
  loginUserFederated.tsx             |     100 |      100 |     100 |     100 |
 src/hooks                           |   23.92 |        0 |    7.14 |   19.72 |
  useAlert.tsx                       |    42.3 |        0 |   11.11 |   36.84 | 34-36,46-74
  useAuth.tsx                        |   40.38 |        0 |   14.28 |   34.04 | 44-66,73,79,85-89,104-169
  useData.tsx                        |       0 |        0 |       0 |       0 | 13-92
 src/utils                           |   75.34 |    63.88 |      70 |   74.24 |
  config.ts                          |     100 |      100 |     100 |     100 |
  configureCognito.ts                |   88.88 |      100 |     100 |    87.5 | 10
  harborRequest.ts                   |   28.57 |        0 |       0 |   30.76 | 36-54
 src/views/SignIn                    |       0 |        0 |       0 |       0 |
  SignIn.tsx                         |       0 |        0 |       0 |       0 | 40-223
```

## To Do

- [x] add unit test coverage for `loginUserFederated`
- [x] fix issue with logout redirect

## Demo

https://github.com/CMS-Enterprise/sbom-harbor-ui/assets/1915925/a52d4c4f-7ec1-41d1-8d21-42f9fcd3c5ce
